### PR TITLE
[BUGFIX] Ensure QP definitions interop with tracked props

### PR DESCRIPTION
**Pull request by pzuraq**
This PR ensures QPs can be defined with native getters/setters or tracked properties. QP observers have been updated to always fire asynchronously, except in a few cases in the router's internals where we flush manually because we need to propagate changes immediately.

### CODE REVIEW

File: packages/@ember/runloop/index.js [hunk 1]
```diff
@@ -26,6 +27,7 @@ let flush;
 if (EMBER_METAL_TRACKED_PROPERTIES) {
   flush = function(queueName, next) {
     if (queueName === 'render' || queueName === _rsvpErrorQueue) {
+      flushSyncObservers();
```
Comment (rwjblue) [inline packages/@ember/runloop/index.js]: Can you add a test specifically for this (e.g. unrelated to query params)?

**Review by rwjblue — DISMISSED** [review]

File: packages/@ember/runloop/index.js (hunk 1 above)
Comment (pzuraq) [reply packages/@ember/runloop/index.js]: Discussed this with @wycats and @chancancode today actually, and it turns out we don't need to do this for this PR. I'm going to open up a followup with it and the tests.

Comment (rwjblue) [reply packages/@ember/runloop/index.js]: Seems good. Though I _do_ think we need to do this specific thing (guarantee that the sync observers are flushed).

**Review by chancancode — APPROVED** [review]

File: packages/@ember/-internals/runtime/lib/mixins/observable.js [hunk 2]
```diff
@@ -362,11 +362,12 @@ export default Mixin.create({
...
     @param {Object} target The target object to invoke
     @param {String|Function} method The method to invoke
+    @param {Boolean} async Whether the observer is async or not
     @return {Observable}
     @public
   */
-  addObserver(key, target, method) {
-    addObserver(this, key, target, method);
+  addObserver(key, target, method, async) {
+    addObserver(this, key, target, method, async);
```
Comment (rwjblue) [inline packages/@ember/-internals/runtime/lib/mixins/observable.js]: We really need a test for this (unrelated to query params)

File: packages/@ember/-internals/routing/lib/system/route.ts [hunk 3]
```diff
@@ -2558,6 +2588,12 @@ Route.reopen(ActionHandler, Evented, {
         }
       }
 
+      // Some QPs have been updated, and those changes need to be propogated
+      // immediately. Eventually, we should work on making this async somehow.
+      if (EMBER_METAL_TRACKED_PROPERTIES && qpUpdated === true) {
+        flushAsyncObservers(false);
```
Comment (rwjblue) [inline packages/@ember/-internals/routing/lib/system/route.ts]: It is not clear to me at all why we need to sync flush here. Can you make an issue for us to track removing this (likely evaluating the specific failing test that caused it to be made sync)?

File: packages/@ember/-internals/runtime/lib/mixins/observable.js [hunk 4]
```diff
@@ -379,11 +380,12 @@ export default Mixin.create({
...
     @param {Object} target The target object to invoke
     @param {String|Function} method The method to invoke
+    @param {Boolean} async Whether the observer is async or not
     @return {Observable}
     @public
   */
-  removeObserver(key, target, method) {
-    removeObserver(this, key, target, method);
+  removeObserver(key, target, method, async) {
+    removeObserver(this, key, target, method, async);
```
Comment (rwjblue) [inline packages/@ember/-internals/runtime/lib/mixins/observable.js]: Needs a specific test, for users of observable (not related to query params).

File: packages/@ember/-internals/routing/lib/system/route.ts (hunk 3 above)
Comment (pzuraq) [reply packages/@ember/-internals/routing/lib/system/route.ts]: For reference, the failing tests were:

```
<LinkTo /> component with query params (routing): the <LinkTo /> component applies activeClass when query params are not changed
{{link-to}} component with query params (routing): the {{link-to}} applies activeClass when query params are not changed
```

**Pull request merged**